### DISCUSSION TITLE
High CPU during Startup with OTel Java Agent

### DISCUSSION THREAD

**Discussion by suman2007** [open-telemetry/opentelemetry-java-instrumentation#10752] [Q&A]
When I instrument my service with Java Agent the CPU utilization gets almost more than double 530% during startup of service and last for 10 minutes, after that it was reduced and became normal. 
Is the behaviour is expected?
How to tune Java Agent to reduce the CPU utilization?

**Comment by steverao** [comment]
> When I instrument my service with Java Agent the CPU utilization gets almost more than double 530% during startup of service and last for 10 minutes, after that it was reduced and became normal. Is the behaviour is expected? How to tune Java Agent to reduce the CPU utilization?

Which Java Agent version you used? Does this phenomenon exist in an application or does it occur when testing many applications?

**Comment by suman2007** [comment]
I am using Java Agent Version 1.32.0.
I have tested with two different applications, and in both applications, the outcome is the same high CPU during startup and persists for 10-14 minutes. After that, the CPU was getting down slowly to 15%.

![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/55128431/aea2190e-31b0-48ca-b7f7-da096e1628b7)

**Reply by steverao** [reply]
According to information you provided, I didn't reproduce same phenomenon with you. I used CPU Profiling tool provided by IDEA to monitor CPU usage. I created a 1 minute profiling data during the startup phase in picture one, and another 1 minute profiling data in picture two 10 minutes later. There is not much difference in CPU utilization between the two periods.
![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/25423978/01b811ca-f312-4ec4-8180-3932016f6f27)
![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/25423978/75e75ce4-174b-4c8e-9086-86928482685f)
I suggest you can use CPU Profiling tool to monitor which part consume so much CPU resource during the startup phase.

**Reply by suman2007** [reply]
Thanks for the update,
I am running in RHEL 7.9 system with the command: java -javaagent:opentelemetry-javaagent-all.jar -Dotel.metrics.exporter=none -Dotel.traces.exporter=none -Dotel.logs.exporter=none -jar application-SNAP-1.jar

How can I do the profile from a Linux terminal?

**Reply by steverao** [reply]
If you only can produce it in Linux environment, maybe you can use [Arthas](https://arthas.aliyun.com/en/doc/profiler.html) to help you do CPU Profiling analysis.

**Reply by suman2007** [reply]
I have tried with the Arthas but didn't find out the actual bottleneck. Mean time I have tried multiple time and observed the Start Time CPU
My service itself is taking 386% of CPU initially
My service + Agent is taking 439% of CPU initially.
My system has 35 cores.
It could be possible that I am running Otel Agent without excluding any library which is not required for my service, e.g. axis2, geode etc for high CPU.
The injection of those libraries could be a reason for the high CPU during starting of the service.
Is it required to exclude unnecessary libraries?
The command I am using...
java -javaagent:opentelemetry-javaagent-all.jar -Dotel.metrics.exporter=none -Dotel.traces.exporter=none -Dotel.logs.exporter=none -jar application-SNAP-1.jar

![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/55128431/9179325e-9678-455e-88d6-74caaad542c9)

**Reply by steverao** [reply]
From your description above, the root cause may come from your own application. For second point you mentioned, If you don't use some framework such as axis2, geode. They will have no effect to your system resource such as CPU, you don't need to exclude them.

**Reply by suman2007** [reply]
Thanks for updating, I have checked and there is a new result below--->
When I used OTel Java instrumentation CPU utilization increased 35% depending on the Application load.

**Reply by steverao** [reply]
The percentage of CPU utilization is hard to describe the overhead. Can you tell us how many cores increase and that is common performance in other application or not?

**Reply by suman2007** [reply]
My system has 46 cores.  There is not any load in the service. CPU% was calculated with  ps -p $pid -o %cpu
                       **Initial CPU%**      
Service:                  401                         
Service+Agent      436                          

                       **After 15mints**
Service:                  14%
Service+Agent       21%

**Reply by steverao** [reply]
Java Agent is used to modify your application's bytecode and add some aspect logics. Its overhead is influenced by many factors, such as the number of classes, QPS, and etc. So at first, if your app has many classed need to initialize, it will take up more resources than usual. I think Java Agent should be normal and you had better to find the root cause why your app takes so much CPU resource when it starts. Hi @laurit, for the problem, WDYT? Whether we have a benchmark report of Java Agent before?

**Reply by suman2007** [reply]
@steverao Absolutely correct, I am sharing a report with a spring boot application with minimum code having only 1 class.

Without Agent  Initial CPU%: (299+294+285)/3 = 293
With Agent  Initial CPU%: (309+298+297)/3 = 301
~ 8%

The query is that if I have 10 services hosted as containers then overall overhead in the Host system will increase by 80% during startup.

**Reply by steverao** [reply]
Use Java Agent will have some extra overhead for each service, I think you can calculate the approximate overall overhead of 10 services according to way you mentioned above and reserve some system resources when you use it.
